app/models/piece.rb

```diff
@@ -8,12 +8,13 @@ def move_to!(new_row, new_col)
     if type == "Pawn" && check_adjacent_pieces(new_row, new_col)
       capture_en_passant!(new_row, new_col, @last_updated) && return
     end
-    update(row_position: new_row, col_position: new_col, moved: true) && return unless @piece
-    if @piece.user_id != user_id
-      @piece.update(row_position: nil, col_position: nil, captured: true)
-      update(row_position: new_row, col_position: new_col, moved: true)
-    else
-      check_if_castling(new_row, new_col)
+    check_if_castling(new_row, new_col) if type == "King"
+    if valid_move?(new_row, new_col) # rubocop:disable Style/GuardClause
```

Comment:
Is there a reason why you insisted on not doing something like
`return unless valid_move?(new_row, new_col)`?